compiler/src/dotty/tools/dotc/core/Denotations.scala

```diff
@@ -449,56 +450,66 @@ object Denotations {
+         *  definitions from different classes that have the same signature but
+         *  conflicting infos. In these cases, do a last minute disambiguation
+         *  by picking one alternative according to the ranking
+         *
+         *    1. Non-bridge methods
+         *    2. non-methods
+         *    3. bridges
+         *    4. NoSymbol
+         *
+         *  If that fails, return a MultiDenotation with both alternatives
+         *  and rely on overloading resolution to pick one of them.
+         */
+        def handleConflict: Denotation =
+
+          def preferMethod(sym1: Symbol, sym2: Symbol): Boolean =
+            sym1.exists &&
+              (!sym2.exists
+              || sym2.is(Bridge) && !sym1.is(Bridge)
+              || sym1.is(Method) && !sym2.is(Method)
+              || sym1.info.isErroneous)
+
+          if preferMethod(sym1, sym2) then denot1
+          else if preferMethod(sym2, sym1) then denot2
+          else
+            overload.println(i"overloaded with same signature: ${sym1.showLocated}: $info1 / ${sym2.showLocated}: $info2")
+            MultiDenotation(denot1, denot2)
+        end handleConflict
 
         if (sym2Accessible && prefer(sym2, sym1, info2, info1)) denot2
-        else {
+        else
           val sym1Accessible = sym1.isAccessibleFrom(pre)
           if (sym1Accessible && prefer(sym1, sym2, info1, info2)) denot1
           else if (sym1Accessible && sym2.exists && !sym2Accessible) denot1
           else if (sym2Accessible && sym1.exists && !sym1Accessible) denot2
-          else if (isDoubleDef(sym1, sym2)) handleDoubleDef
-          else {
-            val sym =
-              if (preferSym(sym2, sym1)) sym2
-              else sym1
-            val jointInfo =
-              try infoMeet(info1, info2, sym1, sym2, safeIntersection)
-              catch {
-                case ex: MergeError =>
-                  // TODO: this picks one type over the other whereas it might be better
-                  // to return a MultiDenotation instead. But doing so would affect lots of
-                  // things, starting with the return type of this method.
-                  if (preferSym(sym2, sym1)) info2
-                  else if (preferSym(sym1, sym2)) info1
-                  else if (pre.widen.classSymbol.is(Scala2x) || migrateTo3)
-                    info1 // follow Scala2 linearization -
-                  // compare with way merge is performed in SymDenotation#computeMembersNamed
-                  else throw new MergeError(ex.sym1, ex.sym2, ex.tp1, ex.tp2, pre)
-              }
-            new JointRefDenotation(sym, jointInfo, denot1.validFor & denot2.validFor, pre)
-          }
-        }
-      }
+          else if isDoubleDef(sym1, sym2) then handleConflict
+          else try
+            val sym = if preferSym(sym2, sym1) then sym2 else sym1
+            val jointInfo = infoMeet(info1, info2, sym1, sym2, safeIntersection)
```

Comment:
`infoMeet` contains code to merge polymorphic and non-polymorphic overloads together, that was necessary to get the standard library to compile, but I think can be removed with the changes in this PR and would be more correct since the current behavior prevents calling one of those overload.